docs/reference/certificates.md

```diff
@@ -0,0 +1,131 @@
+
+A tool like [kubeadm][kubeadm] can be used to generate your certificates for you.
+But there are also reasons to generate these certificates yourself.
+This allows you to keep your private keys off of the api server, for enhanced security.
+
+# Certificates in Kubernetes
+
+Kuberentes uses TLS extensively for authentication.
+
+## Where they're used
+
+Kubernetes uses PKI for the following operations:
+1. Client certificates for the kubelet to authenticate to the API server
+2. Server certificate for the API server endpoint
+3. Client certificates for administrators of the cluster authenticating to the API server
+4. Client certificates for the API server to talk to the kubelets
+5. Client certificate for the API server to talk to etcd
+6. Client certificate/kubeconfig for the controller manager to talk to the API server
+7. Client certificate/kubeconfig for the scheduler to talk to the API server.
+8. Client and server certificates for the [front-proxy][proxy]
+
+
+In addition, etcd uses mutual TLS for authenticating clients and peers (in an HA cluster).
+
+## Where to find them
+
+Kubernetes stores certificates by default in `/etc/kubernetes/pki`. All paths in this documentation are relative to that directory.
+
+## Certificate configuration
+
+There are multiple levels of control for kubernetes certificates. 
+At the basic level, a tool like [kubeadm][kubeadm] can generate all required certificates.
+If that isn't desireable, a single root CA, controlled by an adminstrator, can create multiple _intermediate CAs_, and delegate all further creation to Kubernetes itself. 
+
+
+The CAs needed for this to work are:
+
+| path                   | Default CN                | description                      |
+|------------------------|---------------------------|----------------------------------|
+| ca.crt,key             | kubernetes-ca             | the kubernetes general CA        |
+| etcd/ca.crt,key        | kubernetes-front-proxy-ca | For all ETCD-related functions   |
+| front-proxy-ca.crt,key | etcd-ca                   | for the [front-end proxy][proxy] |
+
+If you don't wish to copy these private keys to your API servers, you can generate all certificates yourself. 
+In that case, you will also need to make all of these:
+
+
+
+[2]: For a liveness probe, if self-hosted
+
+| Default CN                    | Parent CA                 | O (in Subject) | kind   | hosts (SAN)                                 |
+|-------------------------------|---------------------------|----------------|--------|---------------------------------------------|
+| kube-etcd                     | etcd-ca                   |                | server | localhost, 127.0.0.1                        |
+| kube-etcd-peer                | etcd-ca                   |                | peer   | <hostname>, <Host IP>, localhost, 127.0.0.1 |
+| kube-etcd-healthcheck-client  | etcd-ca                   | system:masters | peer   |                                             |
```

Comment:
kind for the kube-etcd-healthcheck-client cert should be client

Reply:
Where will the `kube-etcd-healthcheck-client` CN be used? And it acts as a `system:masters` subject.

```bash
$ kube-apiserver -h
--etcd-cafile string                                      SSL Certificate Authority file used to secure etcd communication.
--etcd-certfile string                                    SSL certification file used to secure etcd communication.
--etcd-compaction-interval duration                       The interval of compaction requests. If 0, the compaction request from apiserver is disabled. (default 5m0s)
--etcd-count-metric-poll-period duration                  Frequency of polling etcd for number of resources per type. 0 disables the metric collection. (default 1m0s)
--etcd-keyfile string                                     SSL key file used to secure etcd communication.
--etcd-prefix string                                      The prefix to prepend to all resource paths in etcd. (default "/registry")
--etcd-servers strings                                    List of etcd servers to connect with (scheme://ip:port), comma separated.
--etcd-servers-overrides strings                          Per-resource etcd servers overrides, comma separated. The individual override format: group/resource#servers, where servers are http://ip:port, semicolon separated.
```

Reply:
It is used by the healthcheck for the etcd static pod deployed by kubeadm.

The `system:masters` subject here is a bit misleading and isn't necessary. Since this cert is created using the etcd CA it does not grant any additional permissions above and beyond any other cert that is created from that CA since we aren't enabling RBAC for etcd (at least yet).